asllib/doc/ASLSemanticsReference.tex

```diff
@@ -11,168 +300,2024 @@
+% % \textbf{\texttt{label}} & \textbf{Description} & \textbf{Chapter/Section} \\
+% % \hline
+% Expressions & Chapter~\ref{chap:eval_expr}\\
+% Left-hand sides of assignments & Chapter~\ref{chap:eval_lexpr}\\
+% Slices & Chapter~\ref{chap:eval_slices}\\
+% Patterns & Chapter~\ref{chap:eval_pattern}\\
+% Local declarations & Chapter~\ref{chap:eval_local_decl}\\
+% Statements & Chapter~\ref{chap:eval_stmt}\\
+% Statement blocks & Chapter~\ref{chap:eval_block}\\
+% \texttt{while} and \texttt{repeat} loops & Section~\ref{sec:SemanticsRule.Loop}\\
+% \texttt{for} loops & Section~\ref{sec:SemanticsRule.For}\\
+% Exception catchers & Chapter~\ref{chap:eval_catchers}\\
+% Subprogram invocations & Chapter~\ref{chap:eval_call}\\
+% Specifications & Chapter~\ref{chap:eval_spec}\\
+% \end{tabular}\\
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Reading guide}
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+The set of rules for each label is further split according to the ``shape'' of the AST node.
+For example, expressions are split into literal expressions, binary operation expressions, etc.
+Each such subset of rules is named accordingly and appears in a section with the same name.
+For example, the rule for literal expressions is named SemanticsRule.Lit\footnote{Sometimes rules are split into cases.
+We then drop the SemanticsRule prefix from each case, for succinctness.
+The same convention applies to helper rules.} and defined in
+\secref{SemanticsRule.Lit}.
+
+Each rule is presented using the following template:
+\begin{itemize}
+\item a Prose paragraph gives the rule in English, and corresponds, as much as possible, to the code of the reference implementation ASLRef given at
+ \href{https://github.com/herd/herdtools7//tree/master/asllib}{/herdtools7/asllib};
+\item one or several Example, which, as much as possible, are also given as regression tests in
+\href{https://github.com/herd/herdtools7//tree/master/asllib/tests/ASLSemanticsReference.t}{/herdtools7/asllib/tests/ASLSemanticsReference.t};
+\item a Code paragraph which gives a verbatim of the corresponding implementation in the interpreter of ASLRef
+\href{https://github.com/herd/herdtools7//tree/master/asllib/Interpreter.ml}{/herdtools7/asllib/Interpreter.ml};
+\item a Formal paragraph, which provides semantic rules that essentially give the same information as the Prose paragraph, for both
+the sequential semantics and the concurrent semantics, in the form of concurrent execution graphs.
+The latter enables building, for each AArch64 instruction,
+the Intrinsic dependencies used by the AArch64 memory model from the ASL code of the instruction~\cite[B2.3.2]{ArmARM}.
+\end{itemize}
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Basic Utility Relations \label{chap:utility}}
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

Comment:
I'm okay with it, but it would create some inconsistency as the rest of the document seems to follow a bottom-up approach.
I am also fine with inverting the order of technical chapters so that we start with the top level and go down until we reach expressions and then the helper relations.
What do you think?